### ISSUE TITLE
Amazon Lambda Test Tool should load function in separate AssemblyLoadContext

### ISSUE THREAD

**Issue by normj** [aws/aws-lambda-dotnet#453]
Currently when the Amazon Lambda Test Tool loads the user specified function it loads the assemblies in the default [AssemblyLoadContext](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.loader.assemblyloadcontext?view=netcore-2.1). This was done to match the behavior of the real Lambda environment.

This can cause problems though since the test tool is ASP.NET Core application which loads up more then just the assemblies specified in Microsoft.NETCore.App like Newtonsoft.Json. If the Lambda function attempts to use a different version then is loaded by Microsoft.AspNetCore.App the function could fail to execute loading the different version.

To solve this the test tool could load the function in a separate AssemblyLoadContext which would allow it to load a separate version of the assemblies like Newtonsoft.Json. 

The .NET Core 1.0 Lambda runtime loaded the provided function in separate AssemblyLoadContext which proved problematic because there were parts of ASP.NET Core that would look for Assemblies in the default AssemblyLoadContext regardless of which context the current object was in. These problems might still exists in current versions and in that case we might want to make it an option to load the function in a separate AssemblyLoadContext.

**Comment by tshannon** [comment]
This was stopping me from doing any debugging on my lambda function because of the Newtonsoft.Json conflict. So I've forked the tool repo and upgraded Newtonsoft.Json to 12.0.2 https://github.com/tshannon/aws-lambda-dotnet

Get the new binaries here https://github.com/tshannon/aws-lambda-dotnet/releases/tag/nj12.0.2

Then change the debug executable in your project properties to point at this.

**Comment by hobojoe** [comment]
@tshannon Tried using your binaries but it also failed :( Same error

**Comment by ebol2000** [comment]
I ran into this as well... related to the linked Newtonsoft.Json issue. I since discovered after trying to deploy the Lambda and run in AWS, that the Newtonsoft issue is not only a local debug tool issue... I ran into the same problem in the Lambda. Only once I reverted all my assemblies (very painfully I might add) back to Newtonsoft.Json 11.0.2 did the Lambdas run. Therefore, I'm thinking the Newtonsoft issue may possibly be a slight different issue than this one.

**Comment by sinexflecs** [comment]
@ebol2000 my lambdas run fine on AWS, just using the debug tool that they fail :(

**Comment by ebol2000** [comment]
@sinexflecs ... but you got Newtonsoft.Json 12.0.2 to run? I got ...
![image](https://user-images.githubusercontent.com/14001759/58342054-9538b000-7e04-11e9-9833-43be8dee8389.png)

Once I reverted all NuGet references back to 11 my issue went away.

**Comment by hobojoe** [comment]
It is a pain to revert as I'm using the new features of Newtonsoft :( 
Would like to see a fix from AWS team...

**Comment by george-weare8** [comment]
Same issue for me only in Lambda tool as far as I can tell so far (thank god!).

`System.IO.FileLoadException: Could not load file or assembly 'Newtonsoft.Json, Version=12.0.0.0, Culture=neutral, PublicKeyToken=30ad4fe6b2a6aeed'. Could not find or load a specific file. (Exception from HRESULT: 0x80131621)`

**Comment by emanuelmiranda83** [comment]
Same issue for me.

**Comment by HalidCisse** [comment]
Same here.

**Comment by preqproaws** [comment]
Same here

**Comment by kevinjpetersen** [comment]
Any updates with this one? @normj

**Comment by bradfuller412** [comment]
I'm running into the same problem. I'm referencing a package that uses newtonsoft.json v12 and I haven't found a workaround that allows me to use the tool. Downgrading is not an option for me unfortunately. Is there any current plan to address this or does anyone have a workaround? Thanks!

**Comment by normj** [comment]
I push out version 0.9.3 of the tool which has a stop gap solution instead of the AssemblyLoadContext solution. I made Amazon.Lambda.TestTool-2.1 explicitly depend on the latest version of Newtonsoft which is 12.0.2. Since .NET Core has a more relaxed loading of assemblies then .NET Framework if the Lambda function was using an older version of Newtonsoft it will allow the process to use the newer version of Newtonsoft. 

The AWS Toolkit for Visual Studio should auto update to 0.9.3 the next time you load a Lambda project. I was detecting an issue in VS 2019 where that didn't happen. I would appreciate feedback on what others notice with the auto update feature in VS 2019. If for some reason it doesn't update then run the following command.

```
dotnet tool update -g Amazon.Lambda.TestTool-2.1
```

**Comment by bradfuller412** [comment]
@normj Thank you for the update, I am now able to use the tool for testing my lambda function! I am using VS 2017 and had to update manually.

**Comment by normj** [comment]
Interesting that VS2017 didn't auto update. Thanks for letting me know.

**Comment by normj** [comment]
I have started working on the .NET Core 3.1 version of this tool for the upcoming .NET Core 3.1 Lambda runtime. The work for that can be tracked in the [mock-testtool-31](https://github.com/aws/aws-lambda-dotnet/tree/mock-testtool-31). In this version I will be loading the Lambda function in a separate AssemblyLoadContext using the new [AssemblyDependencyResolver](https://docs.microsoft.com/en-us/dotnet/core/tutorials/creating-app-with-plugin-support) that was added as part of .NET Core 3.0.
